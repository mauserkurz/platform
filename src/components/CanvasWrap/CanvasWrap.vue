<template>
  <div class="canvas-wrap">
    test
  </div>
</template>

<script>
export default {
  name: 'CanvasWrap',
};
</script>

<style scoped lang="less">
.canvas-wrap {}
</style>
