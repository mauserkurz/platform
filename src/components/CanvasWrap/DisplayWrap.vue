<template>
  <div class="display-wrap">
    <Display/>
  </div>
</template>

<script>
// TODO add logic to store
// TODO store unit test
// TODO unit test
// TODO add es6 - es10 babel plugins
// TODO add exit game
// TODO add restart level
// TODO add restart game
// TODO add pause
// TODO add health
// TODO add monster
// TODO add help with keys and rules
import { mapState, mapMutations, mapActions } from 'vuex';
import GAME_LEVELS from '@/gameLevels';
import Display from '@/components/Display/Display.vue';
import { ARROW_KEY_LIST, STATUS_MAP } from '@/consts';

// TODO add comments
export default {
  name: 'DisplayWrap',

  components: { Display },

  data() {
    return { arrowKeys: {} };
  },

  computed: {
    ...mapState('core', ['status']),
  },

  methods: {
    ...mapMutations('level', ['ADD_END_LEVEL', 'ADD_END_LEVEL']),
    ...mapActions('core', ['start', 'update']),

    trackArrowKeys() {
      const track = (event) => {
        if (this.status !== STATUS_MAP.PLAYING) {
          this.resetArrowKeys();
          return;
        }

        if (ARROW_KEY_LIST.includes(event.key)) {
          this.arrowKeys[event.key] = event.type === 'keydown';
          event.preventDefault();
        }
      };
      window.addEventListener('keydown', track);
      window.addEventListener('keyup', track);
    },

    resetArrowKeys() {
      this.arrowKeys = {};
    },

    runAnimation(frameFunc) {
      const SECOND = 1000;
      const TICK = 100;
      let lastTime = null;

      const frame = (time) => {
        if (lastTime !== null) {
          const timeStep = Math.min(time - lastTime, TICK) / SECOND;

          if (frameFunc(timeStep) === false) {
            return;
          }
        }
        lastTime = time;
        requestAnimationFrame(frame);
      };
      requestAnimationFrame(frame);
    },

    async runLevel(plan) {
      let ending = 1;

      this.start(plan);
      this.ADD_END_LEVEL();

      return new Promise((resolve) => {
        this.runAnimation((time) => {
          this.update({ time, keys: this.arrowKeys });

          if (this.status === STATUS_MAP.PLAYING) {
            return true;
          }
          if (ending > 0) {
            ending -= time;
            return true;
          }
          this.ADD_END_LEVEL();
          resolve(this.status);
          return false;
        });
      });
    },

    async runGame(plans) {
      this.trackArrowKeys();

      for (let level = 0; level < plans.length;) {
        // eslint-disable-next-line no-await-in-loop
        const status = await this.runLevel(plans[level]);

        if (status === STATUS_MAP.WON) {
          level += 1;
        } else if (status === STATUS_MAP.LOST) {
          // TODO replace alert with component notificator
          // eslint-disable-next-line no-alert
          alert('You lost');
        }
        this.resetArrowKeys();
      }
      // TODO replace alert with component notificator
      // eslint-disable-next-line no-alert
      alert("You've won!");
    },
  },

  mounted() {
    this.runGame(GAME_LEVELS);
  },
};
</script>

<style scoped lang="less">
.display-wrap {}
</style>
