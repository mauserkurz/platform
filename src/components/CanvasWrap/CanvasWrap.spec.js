import { mount } from '@vue/test-utils';
import CanvasWrap from '@/components/CanvasWrap/CanvasWrap.vue';

describe('component CanvasWrap', () => {
  describe('rendering', () => {
    it('should render html', () => {
      const wrapper = mount(CanvasWrap);

      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
