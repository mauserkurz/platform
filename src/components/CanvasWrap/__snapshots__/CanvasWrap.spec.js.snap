// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component CanvasWrap rendering should render html 1`] = `
<div class="canvas-wrap">
  test
</div>
`;
